### ISSUE TITLE
[Help] What is the TMDBHelper episode path?

### ISSUE THREAD

**Issue by mathu204** [jurialmunkey/plugin.video.themoviedb.helper#393]
Sorry this is not much of an issue for TMDBH, than a request. Long story short im using an ExtendedInfoMod that lets me play with TMDBHelper, problem is it works for movies but not episodes, so I was trying to see if i could make a player for OpenMeta/Info that sends to TMDBHelper, someone on Reddit helped me out alot and got me started with this player, it works for movies but not episodes, the extendedinfo mod already works with TMDBH without the player so Im trying to figure out what the correct path is for the TMDBH episodes is would u mind letting me know what it is or helping me out it would be greatly appreciated.

this was the original one she gave me and then i tried to change it to this after it didnt work (i dont think it lets me paste two codes in one post the original is in the next comment)

`{
 "name"         : "TMDB Helper",
 "plugin"       : "plugin.video.themoviedb.helper",
 "priority"     : 500,
 "id"           : "direct.themoviedb.helper",
 "movies"       : [[{"link"      : 
 "plugin://plugin.video.themoviedb.helper/?info=play&amp;type=movie&amp;tmdb_id={id}"}]],
 "tvshows"      : [[{"link"      : 
 "plugin://plugin.video.themoviedb.helper?info=play&amp;tmdb_type=tv&amp;tmdb_id={id}&amp;season={season}&amp;episode={episode}"}]]
}`

neither worked for episodes, well the dialog player opens up, but it does not let me play with any of the players i just get errors, i would post the log but its not like its a tmdb helper issue or bug I just need to figure out the right path I assume and then everything is smooth sailing.

**Comment by mathu204** [comment]
`{
 "name"         : "TMDB Helper",

 "plugin"       : "plugin.video.themoviedb.helper",
 "priority"     : 500,
 "id"           : "direct.themoviedb.helper",

 "movies"       : [[{"link"      :"plugin://plugin.video.themoviedb.helper/?info=play&amp;type=movie&amp;tmdb_id={id}"}]],
 "tvshows"      : [[{"link"      :"plugin://plugin.video.themoviedb.helper/?info=play&amp;type=episode&amp;tmdb_id={id}&amp;trakt_id={trakt}&amp;title={name}&amp;season={season}&amp;episode={episode}"}]]
}`

**Comment by drinfernoo** [comment]
The easiest way to find the path to an item: highlight the item, open the context menu, and choose "Add to Favorites". You can then open `special://userdata/favorites.xml` to see the path.

**Comment by mathu204** [comment]
> The easiest way to find the path to an item: highlight the item, open the context menu, and choose "Add to Favorites". You can then open `special://userdata/favorites.xml` to see the path.

I tried doing it like that, but it just gets errors with the addons I send it to so I think something may be missing I’m not sure, would u mind helping me out?

**Comment by drinfernoo** [comment]
You'll want to make sure that the path you take from `favorites.xml` isn't XML-encoded.

**Comment by mathu204** [comment]
Yea I did but I had to change the parts of the episode out for the path so it can be applied to any episode, basically this is what I have right now but it still doesnt work it looks basically exactly as the path in that TMDB Players link u gave me.
` "tvshows"      : [[{"link"      : 
 `"plugin://plugin.video.themoviedb.helper?info=play&amp;tmdb_id={id}&amp;type=episode&amp;season={season}&amp;episode={episode}"}]]``

**Comment by drinfernoo** [comment]
Take out the `&amp`'s.

**Comment by drinfernoo** [comment]
For what it's worth, this sounds like a terrible usage of both of the mentioned addon's... And certainly not an issue for the repo, considering you're making a player for a different add-on entirely. The issues you're having are simply die to player formatting, and not anything with this add-on.

**Comment by mathu204** [comment]
> For what it's worth, this sounds like a terrible usage of both of the mentioned addon's... And certainly not an issue for the repo, considering you're making a player for a different add-on entirely. The issues you're having are simply due to player formatting, and not anything with this add-on.

Yea I know I mentioned that at the top if this is closed cause of that so be it, but since jurial is the one who made tmdb I thought maybe I could get a quick answer one time and be done with it, okay so for example what should ‘?info=play&amp;type=‘ look like? ?info=play&type=, or ?info=play;type=

**Comment by mathu204** [comment]
> > For what it's worth, this sounds like a terrible usage of both of the mentioned addon's... And certainly not an issue for the repo, considering you're making a player for a different add-on entirely. The issues you're having are simply due to player formatting, and not anything with this add-on.
> 
> Yea I know I mentioned that at the top if this is closed cause of that so be it, but since jurial is the one who made tmdb I thought maybe I could get a quick answer one time and be done with it, okay so for example what should ‘?info=play&type=‘ look like? ?info=play&type=, or ?info=play;type=

Also the only reason I am using it like this is. Because I am not skilled enough to replace extended info with tmdb helper as mentioned by jurial in one part of this wiki, and this extended(open)info mod does the job by letting me use tmdb helper inside of it, again the only problem is I can’t use it on episodes for some reason hence I’m trying to make a player just for the episodes

**Comment by jurialmunkey** [comment]
I'd say probably where you are going wrong is `tmdb_id={id}`. 

I'm not all that familar with OpenMeta code but I'd wager a guess that for tvshows it is actually returning TVDb ID *not* the TMDb ID. In which case you'd need `tvdb_id={id}` instead.

Also very possible that it is returning the ID for the episode whereas TMDbHelper needs the ID for the parent show.

Other params TMDbHelper will accept in lieu of a TMDb ID is `query={tvshowtitle}` (you'll need to replace tvshowtitle with whatever keyname OpenMeta uses).

FYI `trakt_id=` is *not* an accepted param. Also identifier params shouldn't be used together. Options are `query=` `tmdb_id=` `tvdb_id=` `imdb_id=` 

With query you can also specify `year=` for tvshow year and `episode_year=` for episode year to be able to ensure you get better results for tvshows where there are two or more shows that have the same name (e.g. The Office, House of Cards, The Goldbergs, Shameless etc.)

**Comment by mathu204** [comment]
> I'd say probably where you are going wrong is `tmdb_id={id}`.
> 
> I'm not all that familar with OpenMeta code but I'd wager a guess that for tvshows it is actually returning TVDb ID _not_ the TMDb ID. In which case you'd need `tvdb_id={id}` instead.
> 
> Also very possible that it is returning the ID for the episode whereas TMDbHelper needs the ID for the parent show.
> 
> Other params TMDbHelper will accept in lieu of a TMDb ID is `query={tvshowtitle}` (you'll need to replace tvshowtitle with whatever keyname OpenMeta uses).
> 
> FYI `trakt_id=` is _not_ an accepted param. Also identifier params shouldn't be used together. Options are `query=` `tmdb_id=` `tvdb_id=` `imdb_id=`
> 
> With query you can also specify `year=` for tvshow year and `episode_year=` for episode year to be able to ensure you get better results for tvshows where there are two or more shows that have the same name (e.g. The Office, House of Cards, The Goldbergs, Shameless etc.)

Hi, thank you and drinferno for getting back to, and helping me out here again i am really nooby with this stuff but i am trying.
my question is so if i used query={whatever openmeta title id is}, and u said with query I can add other things such as year, how would i do that? could I use`query={tvshowid}&amp;year={year}` or `query={tvshowid}&year={year}`

sorry if this is a stupid question as i am someone with little to none code experience.

**Comment by mathu204** [comment]
Could u give me an example of a plugin path using query=? also in the meantime looking at both TMDBHelper player parameters, and Openmetas Ive come up with this 'plugin://plugin.video.themoviedb.helper/?info=play&amp;type=episode&amp;tvdb_id={id}&amp;title={title}&amp;epid={epid}&amp;season={season}&amp;episode={episode}' hopefully this will work

**Comment by mathu204** [comment]
NVM USING THE ABOVE ONE WORKED!! THANK YOU BOTH SO MUCH!!!!

**Comment by jurialmunkey** [comment]
Title= and epid= are not valid params for the play function url. You should remove those.

Only params play url accepts are tmdb_id tvdb_id imdb_id season episode query year episode_year

**Comment by mathu204** [comment]
> Title= and epid= are not valid params for the play function url. You should remove those.
> 
> Only params play url accepts are tmdb_id tvdb_id imdb_id season episode query year episode_year

Alright thank you, I do have one last question if it is possible, can I insert anythign into the code to make it so it always brings up the player dialog rather than using the default players I have set in TMDB?

**Comment by jurialmunkey** [comment]
> Alright thank you, I do have one last question if it is possible, can I insert anythign into the code to make it so it always brings up the player dialog rather than using the default players I have set in TMDB?

Yes you can add `ignore_default=true`

**Comment by mathu204** [comment]
> > Alright thank you, I do have one last question if it is possible, can I insert anythign into the code to make it so it always brings up the player dialog rather than using the default players I have set in TMDB?
> 
> Yes you can add `ignore_default=true`

cool thanks, would i put it as &amp; again or?

**Comment by jurialmunkey** [comment]
> cool thanks, would i put it as & again or?

Yes, parameters in a url query string are separated by ampersands `&`
https://en.wikipedia.org/wiki/Query_string

A question mark separates the base url (in this case `plugin://plugin.video.themoviedb.helper/`) from the parameters that are sent to it. So what you are doing with that url is telling kodi to call the plugin named `plugin.video.themoviedb.helper` and give it the variables listed with that they equal. So you are saying the info is play so run the play code and give it this tmdb_id, this season number, this episode number etc.

**Comment by mathu204** [comment]
Alright, if i used the info=query, would it still play the episode automatically after finding it?

Also, lets say i use info=query, and add parameters for query=tv show title, episode title, year, epyear, etc. If it finds Tv show title, year, and ep year but not the name will it still play? So my question is can i add a bunch of parameters and if atleast 1, few or many of them are found it will play the best match? or do all parameters have to be met for it to play.

Also ignore me saying info=query, reading up a bit again I realize it should be info=play, and query should be used instead of tvdb id or another identifier.

**Comment by mathu204** [comment]
`{
 "name"         : "TMDB Helper",
 "plugin"       : "plugin.video.themoviedb.helper",
 "priority"     : 500,
 "id"           : "direct.themoviedb.helper",
 "movies"       : [[{"link"      : 
 "plugin://plugin.video.themoviedb.helper/?info=play&amp;type=movie&amp;tmdb_id={id}&amp;ignore_default=true"}]],
 "tvshows"      : [[{"link"      : 
 "plugin://plugin.video.themoviedb.helper/?info=play&amp;type=episode&amp;query={clearname}&amp;year={year}&amp;title={title}season={season}&amp;episode={episode}&amp;ignore_default=true"}]]
}
`

sorry to bother you with this but i appreciate the fact that you respond and give me advice with this. I made these parameters, can you tell me if it seems right to you or there is something that is conflicting or i may not need? Im about to test it out now too

**Comment by jurialmunkey** [comment]
So to play an episode we need the TMDb ID `tmdb_id=`, the season number `season=`, and the episode number `episode=`. If we don't have an ID then TMDbHelper will try to look it up using the other params. 

The preference order is (1) tmdb_id; (2) imdb_id; (3) tvdb_id; (4) query. It only moves onto using the next type of lookup if the earlier params are empty (or not specified at all).

Probably the best way to explain is by giving a few different example scenarios:

`info=play&type=tv&tmdb_id=4556&season=9&episode=1&imdb_id=tt0285403&tvdb_id=76156&query=Scrubs&year=2001&episode_year=2009`
This will use the TMDb ID to play S09E01 of Scrubs. Even though I've listed all the possible params here, the other tvdb_id/imdb_id/query/year/episode_year params do nothing. It is fine to specify the other params, they just don't get used.

`&tvdb_id=76156&query=Scrubs`
This will give the TVDb ID "76156" to the TMDb ID find function. The query param is not used and does nothing.

`&tvdb_id=&query=Scrubs`
This will give the query "Scrubs" to the TMDb search function. The TMDb ID for the first search result returned will be used.

`&tvdb_id=Foobar&query=Scrubs`
This will fail because the TVDb ID is invalid and will return no results. The query param is not used and does nothing because a TVDb ID was specified (even though the ID is invalid).

`&tvdb_id=&query=Scrubs&year=2009&season=9&episode=1`
This will fail because even though this episode aired in 2009, Scrubs first aired in 2001. When year is specified, the query result with the matching year is taken. Because we are looking up a tvshow, it needs to be the year of the tvshow not the episode.

`&tvdb_id=&query=Scrubs&episode_year=2009&season=9&episode=1`
This will be successful because Scrubs first aired before 2009. When the episode year is specified, the search results for the tvshow lookup are ordered in year descending order and then the first result where the year is equal to or earlier than the episode year is taken.

**Comment by mathu204** [comment]
Dope, thank you this was really informative. I have an addon that I won’t mention, but when playing directly through that, for example ep 1 of a tv show, I get a lot of content. But through the tmdb helper player for it I barely get any, Ive been messing around with adding different parameters and the closest I got was getting content to show up but say only the first 4 were correct, rest were incorrectly named and so on.

I had previously assumed that by using the tmdb player as long as it would find the show and episode through the correct parameters and within the addon, it would pass on all the relevant information to the player I select, but based on how changing the parameters can affect it a lot it seems like this isn’t the case. I don’t want to go too much into the details since it isn’t an official addon but there is a lot of stuff in the player for that addon that I checked such as plot, fan art, etc. does adding this to the player really make a difference?
